presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.reader;
+
+import com.facebook.presto.memory.context.LocalMemoryContext;
+import com.facebook.presto.orc.StreamDescriptor;
+import com.facebook.presto.orc.TupleDomainFilter;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.Int128ArrayBlock;
+import com.facebook.presto.spi.type.DecimalType;
+import com.facebook.presto.spi.type.UnscaledDecimal128Arithmetic;
+import io.airlift.slice.Slice;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.UnscaledDecimal128Arithmetic.rescale;
+
+public class LongDecimalSelectiveStreamReader
+        extends AbstractDecimalSelectiveStreamReader
+{
+    public LongDecimalSelectiveStreamReader(
+            StreamDescriptor streamDescriptor,
+            Optional<TupleDomainFilter> filter,
+            DecimalType decimalType,
+            boolean outputRequired,
+
+            LocalMemoryContext systemMemoryContext)
+    {
+        super(streamDescriptor, filter, outputRequired, decimalType, systemMemoryContext);
+    }
+
+    @Override
+    int readNoFilter(int[] positions, int positionCount)
+            throws IOException
+    {
+        // filter == null implies outputRequired == true
+        int streamPosition = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skip(position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                if ((nonDeterministicFilter && filter.testNull()) || nullsAllowed) {
+                    if (outputRequired) {
+                        nulls[outputPositionCount] = true;
+                    }
+                    outputPositionCount++;
+                }
+            }
+            else {
+                Slice decimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
```

Comment:
- `unscaledDecimal` allocates new Slice and new byte buffer; these allocations happen for each row, but are not necessary; moving decimal and rescaledDecimal variables outside of the loop and reusing them in each iteration will avoid unnecessary allocations
- rescaledDecimal.getLong performs an unnecessary boundary check twice for each row (4 times for the case with filter); these checks are unnecessary and better avoided; 
- I made an *ugly* sketch of the above changes, re-run the bechmark and saw 1.4x improvement.

```
# Run complete. Total time: 00:02:17

Benchmark                             (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkSelectiveStreamReaders.read   decimal(30,10)         true  avgt   20  0.368 ± 0.027   s/op
BenchmarkSelectiveStreamReaders.read   decimal(30,10)        false  avgt   20  0.585 ± 0.010   s/op



# Run complete. Total time: 00:02:09

Benchmark                             (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkSelectiveStreamReaders.read   decimal(30,10)         true  avgt   20  0.259 ± 0.006   s/op
BenchmarkSelectiveStreamReaders.read   decimal(30,10)        false  avgt   20  0.407 ± 0.008   s/op
```

Here is a sketch of the changes:

```
// no filter

        Slice decimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
        Slice rescaledDecimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
        long address = rescaledDecimal.getAddress();

        // filter == null implies outputRequired == true
        int streamPosition = 0;
        for (int i = 0; i < positionCount; i++) {
            ...
                int scale = (int) scaleStream.next();
                dataStream.nextLongDecimal(decimal);
                rescale(decimal, decimalType.getScale() - scale, rescaledDecimal);
                byte[] data = (byte[]) rescaledDecimal.getBase();
                long low = unsafe.getLong(data, address);
                long high = unsafe.getLong(data, address + Long.BYTES);
                if (outputRequired) {
                    values[2 * outputPositionCount] = low;
                    values[2 * outputPositionCount + 1] = high;
                }
            ...
        }

// filter

        Slice decimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
        Slice rescaledDecimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
        long address = rescaledDecimal.getAddress();

        int streamPosition = 0;
        outputPositionCount = 0;
        for (int i = 0; i < positionCount; i++) {
            ...
            else {
                int scale = (int) scaleStream.next();
                dataStream.nextLongDecimal(decimal);
                rescale(decimal, decimalType.getScale() - scale, rescaledDecimal);
                byte[] data = (byte[]) rescaledDecimal.getBase();
                long low = unsafe.getLong(data, address);
                long high = unsafe.getLong(data, address + Long.BYTES);
                if (filter.testDecimal(low, high)) {
                    if (outputRequired) {
                        values[2 * outputPositionCount] = low;
                        values[2 * outputPositionCount + 1] = high;
                        if (nullsAllowed && presentStream != null) {
                            nulls[outputPositionCount] = false;
                        }
                    }
                    outputPositions[outputPositionCount] = position;
                    outputPositionCount++;
                }
            ...
        }
```